dali/kernels/kernel_manager.h

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_KERNEL_MANAGER_H_
+#define DALI_KERNELS_KERNEL_MANAGER_H_
+
+#include <memory>
+#include <utility>
+#include <functional>
+#include "dali/kernels/scratch.h"
+#include "dali/kernels/context.h"
+#include "dali/kernels/kernel_req.h"
+#include "dali/core/small_vector.h"
+
+namespace dali {
+namespace kernels {
+
+struct AnyKernelInstance {
+  KernelRequirements requirements;
+  std::unique_ptr<void, void(*)(void*)> instance = { nullptr, free };
+
+  template <typename Kernel, typename... Args>
+  Kernel &create_or_get(Args&&... args) {
+    void (*deleter)(void *) = delete_kernel<Kernel>;
+    if (!instance || instance.get_deleter() != deleter) {
+      instance.reset();
+      instance = { new Kernel{std::forward<Args>(args)...}, deleter };
+    }
+    return *static_cast<Kernel*>(instance.get());
+  }
+
+  template <typename Kernel, typename... Args>
+  Kernel &get() {
+    void (*deleter)(void *) = delete_kernel<Kernel>;
+    if (!instance)
+      throw std::logic_error("The kernel instance is null");
+    if (instance.get_deleter() != deleter)
+      throw std::logic_error("The kernel instance is of different type than requested");
+
+    return *static_cast<Kernel*>(instance.get());
+  }
+
+  template <typename Kernel>
+  static void delete_kernel(void *ptr) {
+    delete static_cast<Kernel*>(ptr);
+  }
+
+  explicit operator bool() const noexcept { return static_cast<bool>(instance); }
+};
+
+/// @brief Manages multiple instances of run-time typed kernels
+///
+/// KernelManager provides type erasure for kernels whose type is selected at
+/// run-time. Kernel manager also carries out mundane tasks of keeping
+/// ScratchpadAllocators and reserving memory according to requirements returned
+/// by kernel's Setup method.
+///
+/// A scratchpad allocator is created per-thread with thread indexing supported
+/// explicitly by the caller.
+class DLL_PUBLIC KernelManager {
+ public:
+  /// @brief Creates `num_instances` slots for kernels and `num_threads` scratcapads
+  void Initialize(size_t num_instances, size_t num_threads);
```

Comment:
I'm wondering if we should just create the instances here when calling the initialize.

Reply:
That depends on how we construct them. Current way enables the caller to specify different constructor arguments for each instance.

Reply:
Yes, but at the same time GetInstance is not the clearest way of calling the constructor.